webknossos/script-collection/globalize_floodfill.py

```diff
@@ -0,0 +1,364 @@
+                                )
+                            ]:
+                                continue
+                        if inside_bbox(neighbor_pos):
+                            if cube_data[neighbor_pos] == source_id or (
+                                inside_bbox(
+                                    global_neighbor_pos,
+                                    already_processed_bbox_top_left,
+                                    already_processed_bbox_bottom_right,
+                                )
+                                and cube_data[neighbor_pos] == target_id
+                            ):
+                                seeds_in_curr_bucket.add(neighbor_pos)
+                        else:
+                            bucket_and_seed_pos.append(
+                                get_bucket_pos_and_offset(global_neighbor_pos)
+                            )
+                if dirty_bucket:
+                    wkw_data.write(current_cube, cube_data)
+
+
+def detect_cube_length(layer_path: Path) -> int:
+    if layer_path is not None:
+        with wkw.Dataset.open(str(layer_path)) as dataset:
+            return dataset.header.block_len * dataset.header.file_len
+    raise RuntimeError(
+        f"Failed to detect the cube length (for {layer_path}) because the layer_path is None"
+    )
+
+
+def detect_bbox(layer_path: Path) -> Optional[dict]:
+    # Detect the coarse bounding box of a dataset by iterating
+    # over the WKW cubes
+    def list_files(layer_path: str) -> Iterable[str]:
+        return iglob(path.join(layer_path, "*", "*", "*.wkw"), recursive=True)
+
+    def parse_cube_file_name(filename: str) -> Tuple[int, int, int]:
+        CUBE_REGEX = re.compile(
+            fr"z(\d+){re.escape(os.path.sep)}y(\d+){re.escape(os.path.sep)}x(\d+)(\.wkw)$"
+        )
+        m = CUBE_REGEX.search(filename)
+        if m is not None:
+            return int(m.group(3)), int(m.group(2)), int(m.group(1))
+        raise RuntimeError(f"Failed to parse cube file name from {filename}")
+
+    def list_cubes(layer_path: str) -> Iterable[Tuple[int, int, int]]:
+        return (parse_cube_file_name(f) for f in list_files(layer_path))
+
+    xs, ys, zs = list(zip(*list_cubes(str(layer_path))))
+
+    min_x, min_y, min_z = min(xs), min(ys), min(zs)
+    max_x, max_y, max_z = max(xs), max(ys), max(zs)
+
+    cube_length = detect_cube_length(layer_path)
+    if cube_length is None:
+        return None
+
+    return {
+        "topLeft": [min_x * cube_length, min_y * cube_length, min_z * cube_length],
+        "width": (1 + max_x - min_x) * cube_length,
+        "height": (1 + max_y - min_y) * cube_length,
+        "depth": (1 + max_z - min_z) * cube_length,
+    }
+
+
+def combine_with_fallback_layer(
+    bbox: dict,
+    output_path: Path,
+    volume_data_path: Path,
+    segmentation_layer_path: Path,
+) -> None:
+    with wkw.Dataset.open(str(volume_data_path)) as volume_data_wkw:
```

Comment:
Two things:
- using the new dataset API would be great
- copying the entire segmentation layer first and then applying the volume tracing ad-hoc is probably faster for the most common use cases

But points are not really blocking, though.

Reply:
also this should live on an own module probably.